provider/k8s/client.go

```diff
@@ -0,0 +1,168 @@
+		token:       token,
+		caCert:      caCert,
+	}, nil
+}
+
+// GetIngresses returns all services in the cluster
+func (c *clientImpl) GetIngresses(predicate func(Ingress) bool) ([]Ingress, error) {
+	getURL := c.endpointURL + extentionsEndpoint + defaultIngress
+	request := gorequest.New().Get(getURL)
+	if len(c.token) > 0 {
+		request.Header["Authorization"] = "Bearer " + c.token
+		pool := x509.NewCertPool()
+		pool.AppendCertsFromPEM(c.caCert)
+		c.tls = &tls.Config{RootCAs: pool}
+	}
+	res, body, errs := request.TLSClientConfig(c.tls).EndBytes()
+	if errs != nil {
+		return nil, fmt.Errorf("failed to create request: GET %q : %v", getURL, errs)
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("http error %d GET %q: %q", res.StatusCode, getURL, string(body))
+	}
+
+	var ingressList IngressList
+	if err := json.Unmarshal(body, &ingressList); err != nil {
+		return nil, fmt.Errorf("failed to decode list of ingress resources: %v", err)
+	}
+	ingresses := ingressList.Items[:0]
+	for _, ingress := range ingressList.Items {
+		if predicate(ingress) {
+			ingresses = append(ingresses, ingress)
+		}
+	}
+	return ingresses, nil
+}
+
+// WatchIngresses returns all services in the cluster
+func (c *clientImpl) WatchIngresses(predicate func(Ingress) bool, stopCh <-chan bool) (chan interface{}, chan error, error) {
+	watchCh := make(chan interface{})
+	errCh := make(chan error)
+
+	getURL := c.endpointURL + extentionsEndpoint + defaultIngress + "?watch=true"
+
+	// Make request to Kubernetes API
+	request := gorequest.New().Get(getURL)
+	if len(c.token) > 0 {
+		request.Set("Authorization", "Bearer "+c.token)
+		pool := x509.NewCertPool()
+		pool.AppendCertsFromPEM(c.caCert)
+		c.tls = &tls.Config{RootCAs: pool}
+	}
+	req, err := request.TLSClientConfig(c.tls).MakeRequest()
+	if err != nil {
+		return watchCh, errCh, fmt.Errorf("failed to create request: GET %q : %v", getURL, err)
+	}
+	request.Client.Transport = request.Transport
+	res, err := request.Client.Do(req)
+	if err != nil {
+		return watchCh, errCh, fmt.Errorf("failed to make request: GET %q: %v", getURL, err)
+	}
+
+	shouldStop := safe.New(false)
+
+	go func() {
+		select {
+		case <-stopCh:
+			shouldStop.Set(true)
+			res.Body.Close()
+			return
+		}
+	}()
+
+	go func() {
+		defer close(watchCh)
+		defer close(errCh)
+		for {
+			var ingressList interface{}
+			if err := json.NewDecoder(res.Body).Decode(&ingressList); err != nil {
+				if !shouldStop.Get().(bool) {
+					errCh <- fmt.Errorf("failed to decode list of ingress resources: %v", err)
+				}
+				return
+			}
+
+			watchCh <- ingressList
+		}
+	}()
+	return watchCh, errCh, nil
+}
+
+// GetServices returns all services in the cluster
+func (c *clientImpl) GetServices(namespace string, predicate func(Service) bool) ([]Service, error) {
+	getURL := c.endpointURL + APIEndpoint + "/namespaces/" + namespace + "/services"
+
+	// Make request to Kubernetes API
+	request := gorequest.New().Get(getURL)
```

Comment:
There is some code duplication, but could be refactor afterward so, that's a nit :)